From 1e9af86028f0b477becd9a93f0a18c8bb35d563f Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <tlanger@maxlinear.com>
Date: Thu, 11 Mar 2021 16:31:07 +0100
Subject: [PATCH] PONRTSYS-8683: I2C timing problems under CPU load

The errors reported via the error interrupts are related to the fifo and
its timing, not a problem on the bus.
Do not trigger the i2c_recover_bus in that case, to not make the timing
even worse.
In addition, print the message about timeout in other place before the
recovery is done, to give a better hint why this is done.
---
 drivers/i2c/busses/i2c-lantiq.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/drivers/i2c/busses/i2c-lantiq.c b/drivers/i2c/busses/i2c-lantiq.c
index 44e73876ba5a..c9210a19abbb 100644
--- a/drivers/i2c/busses/i2c-lantiq.c
+++ b/drivers/i2c/busses/i2c-lantiq.c
@@ -531,10 +531,10 @@ static int ltq_i2c_xfer(struct i2c_adapter *adap, struct i2c_msg msgs[],
 			ret = wait_for_completion_interruptible_timeout(
 				&priv->cmd_complete, LTQ_I2C_XFER_TIMEOUT);
 			if (ret == 0) {
-				i2c_recover_bus(adap);
-				ltq_i2c_hw_init(adap);
 				dev_err(priv->dev, "controller timed out 0x%x\n",
 					priv->msg_err);
+				i2c_recover_bus(adap);
+				ltq_i2c_hw_init(adap);
 				ret = -ETIMEDOUT;
 				goto done;
 			} else if (ret < 0)
@@ -555,8 +555,6 @@ static int ltq_i2c_xfer(struct i2c_adapter *adap, struct i2c_msg msgs[],
 						dev_err(priv->dev, " rx underflow\n");
 
 					ret = -EREMOTEIO;
-					i2c_recover_bus(adap);
-					ltq_i2c_hw_init(adap);
 				}
 
 				goto done;
